pypom/page.py

```diff
@@ -107,16 +107,34 @@ def open(self):
         raise UsageError('Set a base URL or URL_TEMPLATE to open this page.')
 
     def wait_for_page_to_load(self):
-        """Wait for the page to load.
+        """
+        Wait for the page to load.
 
         By default the driver will try to wait for any page loads to be
-        complete, however it's not uncommon for it to return early. To address
-        this you can override :py:func:`wait_for_page_to_load` and implement an
-        explicit wait for a condition that evaluates to ``True`` when the page
-        has finished loading.
+        complete, however it's not uncommon for it to return early.
+
+        :py:func:`wait_for_page_to_load` is a template method. Do not override it.
+        Instead, use :py:func:`_before_wait_for_page_load` or :py:func:`_after_wait_for_page_load`
+        for pre- and post- actions.
+
+        Override :py:func:`_page_loaded` and implement a condition that
+        evaluates to ``True`` when the page has finished loading.
 
         :return: The current page object.
         :rtype: :py:class:`Page`
+        """
+        self._before_wait_for_page_load()
```

Comment:
Hi,

if I don't be wrong I think that it should be possible to keep backwards compatibility, maintaining the same semantic for the ``wait_for_page_to_load`` method and creating a new template method just with another name.

For example:
```
def hooks():
    self.before_wait_for_page_load()
    self.wait_for_page_to_load()
    self.after_wait_for_page_load()
    return self
```

This way we should be able to implement this hooks feature with backwards compatibility in mind:
* existing packages implementing ``wait_for_page_to_load`` will work fine because
* if you want to implement after or before hooks you can override the new methods

Do you think it is possible or am I missing something? @davehunt @kimberlythegeek

Reply:
The issue is that consumers of the package will already be calling `open` or `wait_for_page_to_load`, and in many instances will have subclassed and overridden the implementation of `wait_for_page_to_load`. For your suggestion to work, they'd need to add an explicit call to the new `hooks` method. If there is a way to implement this and retain backwards compatibility then that would be ideal, perhaps you could elaborate on your suggestion, in case I've misunderstood.

Reply:
Sorry, I didn't mentioned that the ``open`` method should call the ``hooks`` method and no more ``wait_for_page_to_load``.

Something like that:
```
    def open(self):
        if self.seed_url:
            self.driver_adapter.open(self.seed_url)
            self.hooks()
            return self
raise UsageError('Set a base URL or URL_TEMPLATE to open this page.')
```

instead of the current implementation:
* https://github.com/mozilla/PyPOM/blob/78fb2a92c2796bce23cae32f988f3be764fa168c/pypom/page.py#L105

This way if someone implemented their own ``wait_for_page_to_load`` it should work without having to change anything and they should not be affected by any problem. PyPOM will be in charge of calling the right "template" method inside the ``open`` method instead and it should be transparent for end users. A problem might occurs only if someone overrides the ``open`` method without calling the super method and in such case you will loose the hooks call but I don't know if overriding the ``open`` method is common for PyPOM users.
Probably we won't get any existing failing tests I guess.

I'm just brainstorming, I hope I've understood well and not missing some important points. Let me know if it might work for you please

Reply:
Oh, I got it, you are right!